iree/hal/cts/command_buffer_dispatch_test.h

```diff
@@ -0,0 +1,143 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_HAL_CTS_COMMAND_BUFFER_DISPATCH_TEST_H_
+#define IREE_HAL_CTS_COMMAND_BUFFER_DISPATCH_TEST_H_
+
+#include "iree/base/api.h"
+#include "iree/base/string_view.h"
+#include "iree/hal/api.h"
+#include "iree/hal/cts/cts_test_base.h"
+#include "iree/testing/gtest.h"
+#include "iree/testing/status_matchers.h"
+
+namespace iree {
+namespace hal {
+namespace cts {
+
+class command_buffer_dispatch_test : public CtsTestBase {
+ protected:
+  void PrepareAbsExecutable() {
+    IREE_ASSERT_OK(iree_hal_executable_cache_create(
+        device_, iree_make_cstring_view("default"), &executable_cache_));
+
+    iree_hal_descriptor_set_layout_binding_t descriptor_set_layout_bindings[] =
+        {
+            {0, IREE_HAL_DESCRIPTOR_TYPE_STORAGE_BUFFER},
+            {1, IREE_HAL_DESCRIPTOR_TYPE_STORAGE_BUFFER},
+        };
+    IREE_ASSERT_OK(iree_hal_descriptor_set_layout_create(
+        device_, IREE_HAL_DESCRIPTOR_SET_LAYOUT_USAGE_TYPE_PUSH_ONLY,
+        IREE_ARRAYSIZE(descriptor_set_layout_bindings),
+        descriptor_set_layout_bindings, &descriptor_set_layout_));
+    IREE_ASSERT_OK(iree_hal_executable_layout_create(
+        device_, /*push_constants=*/0, /*set_layout_count=*/1,
+        &descriptor_set_layout_, &executable_layout_));
+
+    iree_hal_executable_spec_t executable_spec;
+    executable_spec.caching_mode =
+        IREE_HAL_EXECUTABLE_CACHING_MODE_ALIAS_PROVIDED_DATA;
+    executable_spec.executable_format =
+        iree_make_cstring_view(get_test_executable_format());
+    executable_spec.executable_data = get_test_executable_data(
+        iree_make_cstring_view("command_buffer_dispatch_test.bin"));
+    executable_spec.executable_layout_count = 1;
+    executable_spec.executable_layouts = &executable_layout_;
+
+    IREE_ASSERT_OK(iree_hal_executable_cache_prepare_executable(
+        executable_cache_, &executable_spec, &executable_));
+  }
+
+  void CleanupExecutable() {
+    iree_hal_executable_release(executable_);
+    iree_hal_executable_layout_release(executable_layout_);
+    iree_hal_descriptor_set_layout_release(descriptor_set_layout_);
+    iree_hal_executable_cache_release(executable_cache_);
+  }
+
+  iree_hal_executable_cache_t* executable_cache_ = NULL;
```

Comment:
remind me to make an iree::hal::ref<> helper now that I have iree_hal_resource_retain/release.